lib/configuration.ex

```diff
@@ -0,0 +1,102 @@
+defmodule Telepoison.Configuration do
+  @moduledoc false
+
+  @default_route_inference_function &Telepoison.URI.infer_route_from_request/1
+
+  defstruct [:infer_route, :ot_attributes]
+
+  alias HTTPoison.Request
+
+  @spec setup(infer_route: (Request.t() -> String.t()), ot_attributes: [{String.t(), String.t()}]) :: :ok
+  def setup(opts \\ []) do
+    Agent.start_link(fn -> set_configuration(opts) end, name: __MODULE__)
+
+    :ok
+  end
+
+  defp set_configuration(opts) do
+    infer_fn =
+      case Keyword.get(opts, :infer_route) do
+        # Unset, return default function
+        nil ->
+          @default_route_inference_function
+
+        infer_fn when is_function(infer_fn, 1) ->
+          infer_fn
+
+        # Set incorrectly
+        _ ->
+          raise RuntimeError, "The configured :infer_route keyword option value must be a function with an arity of 1"
```

Comment:
@MaeIsBad I based my decision mainly on [this](https://elixirforum.com/t/would-you-raise-this-as-an-argumenterror-runtimeerror-or-other/26050) post because I was thinking about configuration as envs.

I now think `ArgumentError` would be more appropriate since this is in response to a call to `setup`, but if we change it to be `Application`-based I think `RuntimeError` becomes more appropriate again...

TL;DR I'll think about it again while refactoring to read the env